File: python/src/chunkshop/bakeoff/gold.py
"""Gold-query loader for bakeoff configs (SC-003).

`BakeoffConfig.gold_queries` is `str | list[GoldQuery]` — a path to a YAML/JSON
file on disk, or an inline list already validated by pydantic. This module
resolves either shape to a concrete `list[GoldQuery]`.
"""
from __future__ import annotations

import json
from pathlib import Path
from typing import Union

import yaml

from chunkshop.bakeoff.config import GoldQuery


def load_gold_queries(spec: Union[str, list[GoldQuery]]) -> list[GoldQuery]:
    """Resolve a `gold_queries` spec to a validated list of `GoldQuery`.

    - `list[GoldQuery]` — returned as-is (already pydantic-validated).
    - `str` — treated as a filesystem path. `.json` parses as JSON array;
      anything else parses as a YAML list. Each element is run through
      `GoldQuery.model_validate`.
    """
    if isinstance(spec, list):
        return spec
    path = Path(spec)
    if not path.exists():
        raise FileNotFoundError(f"gold_queries file not found: {path}")
    text = path.read_text(encoding="utf-8")
    if path.suffix.lower() == ".json":
        raw = json.loads(text)
    else:
        raw = yaml.safe_load(text)
    if not isinstance(raw, list):
        raise ValueError(
            f"gold_queries file must be a YAML/JSON list; got {type(raw).__name__}"
        )
    return [GoldQuery.model_validate(x) for x in raw]

File: rust/chunkshop/src/bakeoff/gold.rs
//! Gold-query loader. Mirrors `python/src/chunkshop/bakeoff/gold.py`.
//! Resolves `gold_queries: str | list[GoldQuery]` to a concrete `Vec<GoldQuery>`.

use std::path::{Path, PathBuf};

use anyhow::{anyhow, Context, Result};

use super::config::{GoldQueriesSpec, GoldQuery};

pub fn load_gold_queries(spec: &GoldQueriesSpec) -> Result<Vec<GoldQuery>> {
    load_gold_queries_with_base(spec, None)
}

/// Resolve `gold_queries` with an optional base directory for relative paths.
/// The CLI passes the bakeoff-YAML's parent so a path like
/// `docs/samples/bakeoff-ntsb/gold-ntsb.yaml` (or `gold-ntsb.yaml`) resolves
/// regardless of the caller's current working directory.
pub fn load_gold_queries_with_base(
    spec: &GoldQueriesSpec,
    base_dir: Option<&Path>,
) -> Result<Vec<GoldQuery>> {
    match spec {
        GoldQueriesSpec::Inline(list) => Ok(list.clone()),
        GoldQueriesSpec::Path(path_str) => {
            let candidate = PathBuf::from(path_str);
            let path: PathBuf = if candidate.is_absolute() {
                candidate
            } else {
                // Try caller-relative first (current cwd), then base-dir-relative.
                if candidate.exists() {
                    candidate
                } else if let Some(base) = base_dir {
                    base.join(&candidate)
                } else {
                    candidate
                }
            };
            if !path.exists() {
                return Err(anyhow!(
                    "gold_queries file not found: {} (tried as cwd-relative and \
                     bakeoff-config-relative)",
                    path.display()
                ));
            }
            let text = std::fs::read_to_string(&path)
                .with_context(|| format!("read gold_queries file {path_str}"))?;
            let raw: serde_json::Value = match path
                .extension()
                .and_then(|s| s.to_str())
                .map(|s| s.to_ascii_lowercase())
                .as_deref()
            {
                Some("json") => serde_json::from_str(&text)?,
                _ => serde_yaml_ng::from_str(&text)?,
            };
            let arr = raw.as_array().ok_or_else(|| {
                anyhow!(
                    "gold_queries file must be a YAML/JSON list; got {}",
                    match &raw {
                        serde_json::Value::Object(_) => "object",
                        serde_json::Value::String(_) => "string",
                        serde_json::Value::Number(_) => "number",
                        serde_json::Value::Bool(_) => "bool",
                        serde_json::Value::Null => "null",
                        serde_json::Value::Array(_) => unreachable!(),
                    }
                )
            })?;
            arr.iter()
                .map(|v| {
                    serde_json::from_value::<GoldQuery>(v.clone())
                        .map_err(|e| anyhow!("gold query parse error: {e}"))
                })
                .collect()
        }
    }
}
